typed/validate.go

```diff
@@ -185,39 +154,30 @@ func (v *validatingObjectWalker) doList(t *schema.List) (errs ValidationErrors)
 	return errs
 }
 
-func (v *validatingObjectWalker) visitMapItems(t *schema.Map, m *value.Map) (errs ValidationErrors) {
-	for i := range m.Items {
-		item := &m.Items[i]
-		pe := fieldpath.PathElement{FieldName: &item.Name}
-
-		if sf, ok := t.FindField(item.Name); ok {
-			v2 := v.prepareDescent(pe, sf.Type)
-			v2.value = item.Value
-			errs = append(errs, v2.validate()...)
-			v.finishDescent(v2)
-		} else {
-			v2 := v.prepareDescent(pe, t.ElementType)
-			v2.value = item.Value
-			if (t.ElementType == schema.TypeRef{}) {
-				errs = append(errs, v2.errorf("field not declared in schema")...)
-			} else {
-				errs = append(errs, v2.validate()...)
-			}
-			v2.doNode()
-			v.finishDescent(v2)
+func (v *validatingObjectWalker) visitMapItems(t *schema.Map, m value.Map) (errs ValidationErrors) {
+	m.Iterate(func(key string, val value.Value) bool {
+		pe := fieldpath.PathElement{FieldName: &key}
+		tr := t.ElementType
+		if sf, ok := t.FindField(key); ok {
+			tr = sf.Type
+		} else if (t.ElementType == schema.TypeRef{}) {
+			errs = append(errs, errorf("field not declared in schema").WithPrefix(pe.String())...)
```

Comment:
The prefix should be added by the caller, not here?

Reply:
Ah it's because you're *not* calling validate here that you need to prefix, otherwise you lose the field name. Worth a comment.